File: apps/cargo-tracker/src/main/java/com/example/cargotracker/routing/domain/model/CarrierMovement.java
package com.example.cargotracker.routing.domain.model;

import com.example.cargotracker.shared.domain.model.Location;

import java.time.LocalDateTime;
import java.util.Objects;

public class CarrierMovement {

    private final Location departureLocation;
    private final Location arrivalLocation;
    private final LocalDateTime departureTime;
    private final LocalDateTime arrivalTime;

    public CarrierMovement(Location departureLocation, Location arrivalLocation,
                           LocalDateTime departureTime, LocalDateTime arrivalTime) {
        this.departureLocation = Objects.requireNonNull(departureLocation, "departureLocation must not be null");
        this.arrivalLocation = Objects.requireNonNull(arrivalLocation, "arrivalLocation must not be null");
        this.departureTime = Objects.requireNonNull(departureTime, "departureTime must not be null");
        this.arrivalTime = Objects.requireNonNull(arrivalTime, "arrivalTime must not be null");
        if (!arrivalTime.isAfter(departureTime)) {
            throw new IllegalArgumentException("arrivalTime must be after departureTime");
        }
    }

    public Location getDepartureLocation() {
        return departureLocation;
    }

    public Location getArrivalLocation() {
        return arrivalLocation;
    }

    public LocalDateTime getDepartureTime() {
        return departureTime;
    }

    public LocalDateTime getArrivalTime() {
        return arrivalTime;
    }
}

File: docs/article/source/haskell/apps/cargo-tracker/src/Cargotracker/Routing/Domain/Model/Value/CarrierMovement.hs
{- | 輸送区間 (IT1 US24)

ある特定船舶の連続した 1 区間 (出発港 → 到着港 + 出発・到着時刻)。
-}
module Cargotracker.Routing.Domain.Model.Value.CarrierMovement
  ( CarrierMovement (..),
  ) where

import Data.Time (UTCTime)

import Cargotracker.Shared.Domain.Common.UnLocode (UnLocode)

data CarrierMovement = CarrierMovement
  { departureLocation :: !UnLocode
  , arrivalLocation :: !UnLocode
  , departureTime :: !UTCTime
  , arrivalTime :: !UTCTime
  }
  deriving stock (Eq, Show)
